### PR TITLE
xeth, rpc: implement eth_estimateGas. Closes #930

### LINKED ISSUE #930: Add eth_estimateGas method
I talked to christian and gavin yesterday and they both agree that we should have a method which can estimate the gas usage of a transaction, without sending and using up the gas.

This method is now described in the specs: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_estimategas

I already started a branch, but with my limited understanding of the EVM its very hard to get the gas used out of it: https://github.com/frozeman/go-ethereum/tree/estimateGas

I basically create the method `eth_estimateGas`:

```js
case "eth_estimateGas":
		args := new(CallArgs)
		if err := json.Unmarshal(req.Params, &args); err != nil {
			return err
		}

		// this need to return the gas used instead of bytes
		v, err := api.xethAtStateNum(-1).Call(args.From, args.To, args.Value.String(), args.Gas.String(), args.GasPrice.String(), args.Data)
		if err != nil {
			return err
		}

                // this also needs to be changed, depending on the return value
		if v == "0x0" {
			*reply = nil
		} else {
			*reply = newHexNum(big.NewInt(v).Bytes())
		}
```

I could trace the EVM down to https://github.com/frozeman/go-ethereum/blob/estimateGas/core/execution.go#L45

But i have no idea how to return the used gas instead of the VM output. @obscuren could you give me a hint?

### PR DESCRIPTION
**Pull request by obscuren**

### CODE REVIEW

File: rpc/api.go [hunk 1]
```diff
@@ -186,16 +186,24 @@ func (api *EthereumApi) GetRequestReply(req *RpcRequest, reply *interface{}) err
...
+		if err != nil {
 			return err
 		}
 
-		v, err := api.xethAtStateNum(args.BlockNumber).Call(args.From, args.To, args.Value.String(), args.Gas.String(), args.GasPrice.String(), args.Data)
+		// TODO unwrap the parent method's ToHex call
+		if len(gas) == 0 {
+			*reply = newHexData([]byte{})
+		} else {
+			*reply = newHexData(gas)
```
Comment (tgerring) [inline rpc/api.go]: For "estimate gas", we should probably output as newHexNum(), since it's a quantity.

Comment (obscuren) [inline rpc/api.go]: :+1:

**Comment by frozeman** [comment]
Just tested your method:

```js
> web3.eth.estimateGas({from: web3.eth.coinbase, to: '0xdfadbef33b3ddb4631583fe41f69704dccb1f4e0', value: web3.toWei(100,'finney')})
1606
```

is that realistic?
because i though normal tx are 21000 gas?

I tried to send a tx using this amount of gas, then i get:
```js
> web3.eth.sendTransaction({from: web3.eth.coinbase, to: '0xdfadbef33b3ddb4631583fe41f69704dccb1f4e0', value: web3.toWei(100,'finney'), gas: 1606})
web3.js:1791 Uncaught Error: Intrinsic gas too low
> web3.eth.sendTransaction({from: web3.eth.coinbase, to: '0xdfadbef33b3ddb4631583fe41f69704dccb1f4e0', value: web3.toWei(100,'finney'), gas: 1607})
web3.js:1791 Uncaught Error: Intrinsic gas too low
```

**Comment by obscuren** [comment]
hm yeah interesting. The call mechanism doesn't go through the whole tx process. I'll fix this

**Comment by frozeman** [comment]
i am testing. The return looks better `22607`.

**Pull request merged**